SI-6951 Refine ClassTag based unapply after inference.

retronym opened the pull request:
Before, we were getting:

```
typed ClassTag.apply[T#Branch](classOf[Tree$Branch]).unapply(<unapply-selector>) <unapply> (Matcher.unapply[Oak.type](<unapply-selector>) <unapply> ((branch @ _))): T#Branch
```

This is problematic for two reasons:
- We don't actually need to use a class tag in
  pattern matching, because after inference the
  substituted formal type of the unapply method
  is a concrete type, Oak.Branch.
- The class tag check delivers a T#Branch, which
  doesn't line up with the required Oak.Branch
  required by unapply.

Now, we use the inferred type argument `Oak.type` to
double check whether the type is in fact uncheckable,
and retain the inferred type.

In addition, we only consider class tags available
through normal implcits, we disable materialization.
See the code comments for an example of an uncheckable
type that for which a classtag can be materialized.

Review by @adriaanm

retronym committed:
SI-6951 Refine ClassTag based unapply after inference.

Before, we were getting:

    typed ClassTag.apply[T#Branch](classOf[Tree$Branch]).unapply(<unapply-selector>) <unapply> (Matcher.unapply[Oak.type](<unapply-selector>) <unapply> ((branch @ _))): T#Branch

This is problematic for two reasons:

  - We don't actually need to use a class tag in
    pattern matching, because after inference the
    substituted formal type of the unapply method
    is a concrete type, Oak.Branch.
  - The class tag check delivers a T#Branch, which
    doesn't line up with the required Oak.Branch
    required by unapply.

Now, we use the inferred type argument `Oak.type` to
double check whether the type is in fact uncheckable,
and retain the inferred type.

In addition, we only consider class tags available
through normal implcits, we disable materialization.
See the code comments for an example of an uncheckable
type that for which a classtag can be materialized.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2079/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2079/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1366/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1366/

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -3478,11 +3479,26 @@ trait Typers extends Modes with Adaptations with Tags {
           // also skip if we already wrapped a classtag extractor (so we don't keep doing that forever)
-          if (uncheckedTypeExtractor.isEmpty || fun1.symbol.owner.isNonBottomSubClass(ClassTagClass)) unapply
-          else wrapClassTagUnapply(unapply, uncheckedTypeExtractor.get, unappType.paramTypes.head)
+          val isClassTagMember = fun1.symbol.owner isNonBottomSubClass ClassTagClass
```
make this an `@inline def` so we only do this check when necessary?

gkossakowski replied:
Why `@inline` in this case?

adriaanm replied:
i don't think making isClassTagMember a method will suddenly make the enclosing method inlineable, so at this point we should probably avoid calling small methods?

gkossakowski replied:
Care to elaborate? Enclosing method here is `doTypedUnapply` which is very large method and I don't see why it should be inlineable.

Calling small methods that do not allocate objects is totally fine and recommended practice. JIT will take care of that very easily.

adriaanm replied:
exactly, it need not be inlineable so increasing its size is not a problem, thus it's okay to inline this small method -- which is what we did (manually) before this PR

gkossakowski replied:
Ok, now I get it. I'd say that our general rule would be that you force inlining with `@inline` if you have a good reason for that and not that it just doesn't hurt. I don't see it here and that's why I asked.

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -3478,11 +3479,26 @@ trait Typers extends Modes with Adaptations with Tags {
+            val inferredScrutineeType: Option[Type] = unappType.paramTypes.headOption.map { tp =>
+              val applied = treeInfo.dissectApplied(fun1)
+              val unapplyFunTpe = applied.core.tpe
+              unapplyFunTpe.paramTypes.head.subst(unapplyFunTpe.typeParams, applied.targs.map(_.tpe))
```
calling `sym.tpe` considered harmful when `sym` may be a type constructor symbol

adriaanm replied:
didn't see it was a tree -- my bad

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -3437,12 +3437,13 @@ trait Typers extends Modes with Adaptations with Tags {
-        else None
 
       if (!isApplicableSafe(Nil, unappType, List(pt), WildcardType)) {
+        val uncheckedTypeExtractor =
```
be good to add a comment that this `uncheckedTypeExtractor` should imply the `uncheckedTypeExtractor` on line 3485
(to make the canRemedy sound with respect to what we will actually remedy)

retronym replied:
Do you believe that is that actually the case? I couldn't convince myself one way or the other. I see how we can miss an unchecked warning if it isn't. Test case suggestions very welcome...

adriaanm replied:
I think it's pretty close as it stands: `Forall x. (if(x.nonEmpty) foo(x)).isEmpty => x.map(f).flatMap(foo).isEmpty` as long as `f` doesn't influence the emptiness of `foo`

adriaanm commented:
ping @retronym -- I think this PR is almost there: changing `.tpe` to `.typeConstructor` is the only major piece of feedback

retronym committed:
SI-6951 Add test to show we aren't missing .tpeHK

We're getting the inferred type arguments to unapply
from the a tree that we decompose with `Applied`,
rather than from symbols. As such, the call to .tpe
is correct. In the additional test case, the substition
of `M => Array` works.

retronym committed:
SI-6951 Allow for checkable type without base classes

If we ignore an unchecked warning and proceed into the
pattern matcher, the checkable type might be completely
abstract, and as such have no base classes. Use `Any`
in that case.

This wasn't the reported bug, but something I found while
testing its fix.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1652/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2383/

retronym commented:
@adriaanm We're using `Tree#tpe` here, rather than `Symbol#{tpe, tpeHK}`. That's admittedly a bit weird, but was the only way I could figure out to get the inferred type arguments of the `unapply` call. I've added a test to show that is works in the presence of type params of kind `*->*`, and fixed a related bug I find in the process.

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1652/ <br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2383/ <br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

adriaanm commented:
closing to keep the queue in check -- please reopen when ready

```
/localhome/jenkins/c/workspace/pr-scala-testsuite-linux-opt/test/files/run/t6951c.scala [FAILED]
```
